spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveMergeIntoTableReferences.scala

```
@@ -0,0 +1,152 @@
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.catalyst.plans.logical.UnresolvedMergeIntoIcebergTable
+import org.apache.spark.sql.catalyst.plans.logical.UpdateAction
+import org.apache.spark.sql.catalyst.plans.logical.UpdateStarAction
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * A resolution rule similar to ResolveReferences in Spark but handles Iceberg MERGE operations.
+ */
+case class ResolveMergeIntoTableReferences(spark: SparkSession) extends Rule[LogicalPlan] {
+
+  private lazy val analyzer: Analyzer = spark.sessionState.analyzer
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
+    case m @ UnresolvedMergeIntoIcebergTable(targetTable, sourceTable, context)
+        if targetTable.resolved && sourceTable.resolved && m.duplicateResolved =>
+
+      val resolvedMatchedActions = context.matchedActions.map {
+        case DeleteAction(cond) =>
+          val resolvedCond = cond.map(resolveCond("DELETE", _, m))
+          DeleteAction(resolvedCond)
+
+        case UpdateAction(cond, assignments) =>
+          val resolvedCond = cond.map(resolveCond("UPDATE", _, m))
+          // the update action can access columns from both target and source tables
+          val resolvedAssignments = resolveAssignments(assignments, m, resolveValuesWithSourceOnly = false)
+          UpdateAction(resolvedCond, resolvedAssignments)
+
+        case UpdateStarAction(updateCondition) =>
+          val resolvedUpdateCondition = updateCondition.map(resolveCond("UPDATE", _, m))
+          val assignments = targetTable.output.map { attr =>
+            Assignment(attr, UnresolvedAttribute(Seq(attr.name)))
+          }
+          // for UPDATE *, the value must be from the source table
+          val resolvedAssignments = resolveAssignments(assignments, m, resolveValuesWithSourceOnly = true)
+          UpdateAction(resolvedUpdateCondition, resolvedAssignments)
+
+        case o => o
```

This is an unsupported action, right? Why just pass it through?

---

This matches what Spark does, but since we use our own plan nodes we may want to fail sooner. I'm guessing that with our own merge node, we'd get a generic "not resolved" message instead of throwing a better one here?

---

Yeah, we can throw an exception early here.

---

Added an exception.